octobot_commons/async_job.py

```diff
@@ -0,0 +1,185 @@
+# pylint: disable=W0703,R0902,R0913
+#  Drakkar-Software OctoBot-Commons
+#  Copyright (c) Drakkar-Software, All rights reserved.
+#
+#  This library is free software; you can redistribute it and/or
+#  modify it under the terms of the GNU Lesser General Public
+#  License as published by the Free Software Foundation; either
+#  version 3.0 of the License, or (at your option) any later version.
+#
+#  This library is distributed in the hope that it will be useful,
+#  but WITHOUT ANY WARRANTY; without even the implied warranty of
+#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+#  Lesser General Public License for more details.
+#
+#  You should have received a copy of the GNU Lesser General Public
+#  License along with this library.
+import asyncio
+
+import time
+
+from octobot_commons.logging.logging_util import get_logger
+
+
+class AsyncJob:
+    """
+    Async job management
+    """
+
+    NO_DELAY = 0.1
+    DEPENDENCIES_WAIT_TIMEOUT = 300
+
+    def __init__(
+        self,
+        callback,
+        execution_interval_delay=NO_DELAY,
+        min_execution_delay=NO_DELAY,
+        is_periodic=True,
+    ):
+        self.logger = get_logger(self.__class__.__name__)
+        self.callback = callback
+        self.is_started = False
+        self.should_stop = False
+        self.is_periodic = is_periodic
+
+        self.last_execution_time = 0
+        self.execution_interval_delay = execution_interval_delay
+        self.min_execution_delay = min_execution_delay
+
+        self.job_dependencies = []
+        self.running_task_event = asyncio.Event()
+        self.running_task_event.set()
+
+        self.job_task = None
+        self.job_periodic_task = None
+
+    async def run(self, force=False, wait_for_task_execution=False, **kwargs):
+        """
+        Run the job if possible
+        Reschedule the jab in the end
+        :param force: When True, force the execution of the job
+        :param wait_for_task_execution: When True, await running_task_event
+        """
+        if self._should_run_job(force=force, ignore_dependencies=True):
+            if wait_for_task_execution:
+                await self._run_task_as_soon_as_possible(force=force, **kwargs)
+            else:
+                self.job_task = asyncio.create_task(
+                    self._run_task_as_soon_as_possible(force=force, **kwargs)
+                )
+        if not self.is_started and self.is_periodic:
+            self.job_periodic_task = asyncio.create_task(
+                self._run_periodic_task(**kwargs)
+            )
+
+    async def _run_periodic_task(self, **kwargs):
+        """
+        Calls _run() periodically until self.should_stop == True or cancellation
+        """
+        while not self.should_stop:
+            self.is_started = True
+            await asyncio.sleep(self.execution_interval_delay)
```

Comment:
To be able to handle manually triggered updates when `_run_periodic_task` is already runnning, this sleep should either adapt to the last (manual) call or this task be cancelled and rescheduled right after any manual call (to ensure the right sleep time after any update: manual or periodic)